### PR TITLE
7: browser views

### LINKED ISSUE #7: Browser-based info views (settings)
## Acceptance Criteria
* Should be able to view FAQs at lockbox.firefox.com/faqs.html
* Should be able to provide feedback at https://www.surveygizmo.com/s3/4713557/Provide-feedback-for-Firefox-Lockbox-Android
* Should open in-app (custom Chrome tab) and not in the default browser like the website links

### Visual Design

**InVision**
FAQ: https://mozilla.invisionapp.com/share/3SN35UHWJDX#/309979891_FAQ
Provide Feedback: https://mozilla.invisionapp.com/share/3SN35UHWJDX#/309979890_Provide_Feedback

**Zeplin**
FAQ: https://zpl.io/aRYj40K
Provide Feedback: https://zpl.io/2pOY3j8

### PR DESCRIPTION
**Pull request by eliserichards**
Fixes #7 

## Testing and Review Notes
FAQ page opens to https://lockbox.firefox.com/faq.html
Provide feedback opens to https://www.surveygizmo.com/s3/4713557/Provide-feedback-for-Firefox-Lockbox-Android

## Screenshots or Videos
Send feedback survey:
image

FAQ link to documentation:
image

Zeplin:
image

## To Do
Make FAQ page backable (to the itemlist?). Right now it opens a web page, and when the device's back button is clicked it goes to a blank page inside the app (the faq fragment)

- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
- [x] request the "UX" team perform a design review (if/when applicable)
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- [x] check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI
-- added labels to the routes, but didn't scan the pages (since they are temporary links)

### CODE REVIEW

**Comment by eliserichards** [comment]
@sashei I had a whole bunch of local changes from last week that I didn't push, sorry!! Can you take another look? The fragments are no longer necessary in my ccurrent version.

**Comment by sashei** [comment]
@eliserichards I took a quick look; one immediate issue that jumps out at me is that you are opening these websites outside of the app context, IIUC from the designs, we should be displaying them in the app context rather than in the user's preferred browser.

File: app/src/main/java/mozilla/lockbox/action/RouteAction.kt [hunk 1]
```diff
@@ -19,6 +19,7 @@ sealed class RouteAction : Action {
     object Back : RouteAction()
     object LockScreen : RouteAction()
     object Filter : RouteAction()
+    object FaqList : RouteAction()
```
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/action/RouteAction.kt]: I have an alternative suggestion: 

```kotlin
data class ExternalWebsite(val url: String) : RouteAction()
object FAQList : ExternalWebsite(Constant.fxa.faq)
object SendFeedback : ExternalWebsite(Constant.fxa.feedback)
```

The route presenter doesn't need to know about subclasses of `ExternalWebsite`, and the presenters sending the actions don't have to worry about URLs.

File: app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt [hunk 2]
```diff
@@ -109,6 +110,8 @@ class RoutePresenter(
             is RouteAction.Login -> navigateToFragment(action, R.id.fragment_fxa_login)
             is RouteAction.ItemList -> navigateToFragment(action, R.id.fragment_item_list)
             is RouteAction.SettingList -> navigateToFragment(action, R.id.fragment_setting)
+            is RouteAction.FaqList -> openWebsite(Constant.Faq.uri)
+            is RouteAction.SendFeedback -> openWebsite(Constant.FeedbackLink.uri)
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: I spoke with Sasha today and we didn't end up implementing the custom chrome tabs, just passing the correct url via a custom route action (similar to the way you send the id to the item detail view). 

I'm curious to see what you think, and if you would like to proceed in the custom chrome tab direction, or maybe add that as an enhancement in our backlog?

File: app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt [hunk 3]
```diff
@@ -105,29 +108,50 @@ class RoutePresenter(
...
         when (action) {
-            is RouteAction.Welcome -> navigateToFragment(action, R.id.fragment_welcome)
-            is RouteAction.Login -> navigateToFragment(action, R.id.fragment_fxa_login)
+            is RouteAction.Welcome -> navigateToFragment(action, R.id.fragment_item_list)
+            is RouteAction.Login -> navigateToFragment(action, R.id.fragment_item_list)
             is RouteAction.ItemList -> navigateToFragment(action, R.id.fragment_item_list)
             is RouteAction.SettingList -> navigateToFragment(action, R.id.fragment_setting)
             is RouteAction.AccountSetting -> navigateToFragment(action, R.id.fragment_account_setting)
-            is RouteAction.LockScreen -> navigateToFragment(action, R.id.fragment_locked)
+            is RouteAction.LockScreen -> navigateToFragment(action, R.id.fragment_item_list)
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: whoops! remnant of debugging when I was having issues with FxA :)

File: app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt [hunk 4]
```diff
@@ -105,29 +108,50 @@ class RoutePresenter(
 
     private fun route(action: RouteAction) {
         when (action) {
-            is RouteAction.Welcome -> navigateToFragment(action, R.id.fragment_welcome)
-            is RouteAction.Login -> navigateToFragment(action, R.id.fragment_fxa_login)
+            is RouteAction.Welcome -> navigateToFragment(action, R.id.fragment_item_list)
+            is RouteAction.Login -> navigateToFragment(action, R.id.fragment_item_list)
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt]: again, remnant of debugging when I was having issues with FxA :) I had to revert to an earlier commit during my impl and I guess I removed my removal of this 😅

File: app/src/main/java/mozilla/lockbox/action/RouteAction.kt [hunk 5]
```diff
@@ -31,10 +33,18 @@ sealed class RouteAction : Action {
...
-    sealed class DialogFragment(@StringRes val dialogTitle: Int, @StringRes val dialogSubtitle: Int? = null) : RouteAction() {
+    sealed class DialogFragment(
+        @StringRes val dialogTitle: Int,
+        @StringRes val dialogSubtitle: Int? = null
+    ) : RouteAction() {
         class FingerprintDialog(@StringRes title: Int, @StringRes subtitle: Int? = null) :
             DialogFragment(dialogTitle = title, dialogSubtitle = subtitle)
     }
+
+    sealed class MenuItem(val url: String? = null) : RouteAction() {
```
[the remark this answers is not in the document]
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/action/RouteAction.kt]: Good suggestion. I'll go with `AppWebPage` instead. I like that much better.

File: app/src/main/java/mozilla/lockbox/action/RouteAction.kt [hunk 6]
```diff
@@ -46,7 +46,7 @@ sealed class RouteAction(
     sealed class AppWebPage(
         val url: String? = null,
         @StringRes val title: Int? = null
-    ) : RouteAction() {
+    ) : RouteAction(TelemetryEventMethod.tap, TelemetryEventObject.app_web_page) {
```
Comment (linuxwolf) [inline app/src/main/java/mozilla/lockbox/action/RouteAction.kt]: The [telemetry](https://github.com/mozilla-lockbox/lockbox-android/blob/master/docs/metrics.md#list-of-proposed-events) docs calls for these to be more specific (e.g., "show faq" vs "opened web app page").

Instead, consider something like:

```
seald class AppWebPage(
  val url: String? = null,
  @StringRes val title: Int? = null,
  eventObject: TelemetryEventObject
) : RouteAction(TelemetryEventMethod.show, eventObject) {
  object FaqList  : AppWebPage(Constant.Faq.uri, R.string.nav_menu_faq, TelemetryEventMethod.settings_faq)
  object SendFeedback : AppWebPage(Constant.SendFeedback.uri, R.string.nav_menu_feedback, TelemetryEventObject.settings_provide_feedback)
}
```

File: app/src/main/java/mozilla/lockbox/action/RouteAction.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (jhugman) [reply app/src/main/java/mozilla/lockbox/action/RouteAction.kt]: :+1:

File: app/src/main/java/mozilla/lockbox/action/RouteAction.kt (hunk 6 above)
Comment (eliserichards) [reply app/src/main/java/mozilla/lockbox/action/RouteAction.kt]: awesome! thank you!

File: app/src/main/java/mozilla/lockbox/support/Constant.kt [hunk 7]
```diff
@@ -25,6 +25,14 @@ object Constant {
         val scopes = arrayOf(profileScope, lockboxScope, oldSyncScope)
     }
 
+    object Faq {
+        const val uri = "https://mozilla-lockbox.github.io/lockbox-android/"
```
Comment (linuxwolf) [inline app/src/main/java/mozilla/lockbox/support/Constant.kt]: This should point to "https://lockbox.firefox.com/faq.html".

CC @mozilla-lockbox/product and @mozilla-lockbox/ux to confirm the link.

File: app/src/main/res/navigation/graph_main.xml [hunk 8]
```diff
@@ -6,130 +6,150 @@
...
+            android:id="@+id/fragment_webview"
+            android:name="mozilla.lockbox.view.AppWebPageFragment"
+            tools:layout="@layout/fragment_webview">
+        <argument
+                android:name="url"
+                android:defaultValue="@string/missing_url"
+                app:argType="string"/>
+        <argument
+                android:name="title"
+                android:defaultValue="0"
```
Comment (linuxwolf) [inline app/src/main/res/navigation/graph_main.xml]: should this be `@string/missing_title`?

**Comment by devinreams** [comment]
Yay! Tests are passing, UX and code review approvals, and will wait to merge after #272 is merged to master and any conflicts are resolved. 🤘

**Pull request merged**